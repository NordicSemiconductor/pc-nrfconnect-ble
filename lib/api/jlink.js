/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

const familyDefinitions = {
    NRF51: 'nrf51',
    NRF52: 'nrf52',
    UNKNOWN: 'unknown',
};

const deviceTypeDefinitions = {
    NRF51xxx_xxAA_REV1: 'nRF51',
    NRF51xxx_xxAA_REV2: 'nRF51',
    NRF51xxx_xxAC_REV3: 'nRF51',
    NRF51xxx_xxAA_REV3: 'nRF51',
    NRF51xxx_xxAB_REV3: 'nRF51',
    NRF51802_xxAA_REV3: 'nRF51802',
    NRF52832_xxAA_ENGA: 'nRF52832',
    NRF52832_xxAA_REV1: 'nRF52832',
    NRF52832_xxAA_REV2: 'nRF52832',
    NRF52832_xxAA_ENGB: 'nRF52832',
    NRF52840_xxAA_ENGA: 'nRF52840',
    NRF52832_xxAA_FUTURE: 'nRF52832',
    NRF52840_xxAA_FUTURE: 'nRF52840',
    NRF52810_xxAA_REV1: 'nRF52810',
    NRF52810_xxAA_FUTURE: 'nRF52810',
    NRF52832_xxAB_REV1: 'nRF52832',
    NRF52832_xxAB_FUTURE: 'nRF52832',
    NRF51801_xxAB_REV3: 'nRF51801',
};

export const getJlinkDeviceInfo = ({ jlink }) => ({
    family: familyDefinitions[jlink.deviceFamily] || 'unknown',
    type: deviceTypeDefinitions[jlink.deviceVersion] || 'unknown',
    firmwareString: jlink.jlinkObFirmwareVersion,
});
