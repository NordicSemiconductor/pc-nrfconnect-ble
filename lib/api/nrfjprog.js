/* Copyright (c) 2015 - 2017, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Use in source and binary forms, redistribution in binary form only, with
 * or without modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions in binary form, except as embedded into a Nordic
 *    Semiconductor ASA integrated circuit in a product or a software update for
 *    such product, must reproduce the above copyright notice, this list of
 *    conditions and the following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 2. Neither the name of Nordic Semiconductor ASA nor the names of its
 *    contributors may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 * 3. This software, with or without modification, must only be used with a Nordic
 *    Semiconductor ASA integrated circuit.
 *
 * 4. Any software provided in binary form under this license must not be reverse
 *    engineered, decompiled, modified and/or disassembled.
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

const familyDefinitions = {
    NRF51: 'nrf51',
    NRF52: 'nrf52',
    UNKNOWN: 'unknown',
};

const deviceTypeDefinitions = {
    NRF51xxx_xxAA_REV1: 'nRF51',
    NRF51xxx_xxAA_REV2: 'nRF51',
    NRF51xxx_xxAC_REV3: 'nRF51',
    NRF51xxx_xxAA_REV3: 'nRF51',
    NRF51xxx_xxAB_REV3: 'nRF51',
    NRF51802_xxAA_REV3: 'nRF51802',
    NRF52832_xxAA_ENGA: 'nRF52832',
    NRF52832_xxAA_REV1: 'nRF52832',
    NRF52832_xxAA_ENGB: 'nRF52832',
    NRF52840_xxAA_ENGA: 'nRF52840',
    NRF52832_xxAA_FUTURE: 'nRF52832',
    NRF52840_xxAA_FUTURE: 'nRF52840',
    NRF52810_xxAA_REV1: 'nRF52810',
    NRF52810_xxAA_FUTURE: 'nRF52810',
    NRF52832_xxAB_REV1: 'nRF52832',
    NRF52832_xxAB_FUTURE: 'nRF52832',
    NRF51801_xxAB_REV3: 'nRF51801',
};

/**
 * Get J-Link device information for the given serial number. The returned
 * promise resolves with an object on the form:
 * {
 *   family: 'nrf52',
 *   type: 'nRF52832',
 *   firmwareString: 'J-Link OB-SAM3U128-V2-NordicSemi...',
 * }
 *
 * @param {Number|String} serialNumber The serial number to get info for.
 * @returns {Promise<Object>} Resolves with object as described above.
 */
// eslint-disable-next-line import/prefer-default-export
export async function getJlinkDeviceInfo({ jlink }) {
    return new Promise(resolve => {
        resolve({
            family: familyDefinitions[jlink.deviceFamily] || 'unknown',
            type: deviceTypeDefinitions[jlink.deviceVersion] || 'unknown',
            firmwareString: jlink.jlinkObFirmwareVersion,
        });
    });
}
