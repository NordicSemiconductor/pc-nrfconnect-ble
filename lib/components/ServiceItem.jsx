/* Copyright (c) 2016, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 *   1. Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 *   2. Redistributions in binary form, except as embedded into a Nordic
 *   Semiconductor ASA integrated circuit in a product or a software update for
 *   such product, must reproduce the above copyright notice, this list of
 *   conditions and the following disclaimer in the documentation and/or other
 *   materials provided with the distribution.
 *
 *   3. Neither the name of Nordic Semiconductor ASA nor the names of its
 *   contributors may be used to endorse or promote products derived from this
 *   software without specific prior written permission.
 *
 *   4. This software, with or without modification, must only be used with a
 *   Nordic Semiconductor ASA integrated circuit.
 *
 *   5. Any software provided in binary form under this license must not be
 *   reverse engineered, decompiled, modified and/or disassembled.
 *
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

'use strict';

import React, { PropTypes } from 'react';

import AttributeItem from './AttributeItem';
import CharacteristicItem from './CharacteristicItem';

class ServiceItem extends AttributeItem {
    constructor(props) {
        super(props);
        this.bars = 1;
        this.attributeType = 'service';
        this.childAttributeType = 'characteristic';
    }

    renderContent(children) {
        const { button } = this.props;

        return (
            <div className="content">
                {this.renderName()}
                {button &&
                    <button className="btn btn-primary btn-nordic" onClick={() => button.onClick()}>
                        <img src={button.icon} alt="" />
                    </button>
                }
                {children}
            </div>
        );
    }

    renderChildren() {
        const {
            item,
            selected,
            addNew,
            selectOnClick,
            onAddDescriptor,
            onSelectAttribute,
            onSetAttributeExpanded,
            onReadCharacteristic,
            onWriteCharacteristic,
            onReadDescriptor,
            onWriteDescriptor,
        } = this.props;

        const {
            children,
        } = item;

        return children.map(characteristic =>
            <CharacteristicItem
                key={characteristic.instanceId}
                item={characteristic}
                selectOnClick={selectOnClick}
                selected={selected}
                onSelectAttribute={onSelectAttribute}
                onSetAttributeExpanded={onSetAttributeExpanded}
                onRead={onReadCharacteristic}
                onWrite={onWriteCharacteristic}
                onReadDescriptor={onReadDescriptor}
                onWriteDescriptor={onWriteDescriptor}
                onChange={() => this.LchildChanged()}
                addNew={addNew}
                onAddDescriptor={onAddDescriptor}
            />,
        );
    }
}

ServiceItem.propTypes = {
    item: PropTypes.object.isRequired,
    selected: PropTypes.string,
    button: PropTypes.shape({
        icon: PropTypes.string,
        onClick: PropTypes.func,
    }),
    onAddDescriptor: PropTypes.func,
    onSelectAttribute: PropTypes.func,
    onSetAttributeExpanded: PropTypes.func,
    onReadCharacteristic: PropTypes.func,
    onWriteCharacteristic: PropTypes.func,
    onReadDescriptor: PropTypes.func,
    onWriteDescriptor: PropTypes.func,
};

export default ServiceItem;
