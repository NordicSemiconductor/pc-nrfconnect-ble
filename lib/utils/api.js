/* Copyright (c) 2016, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 *   1. Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 *   2. Redistributions in binary form, except as embedded into a Nordic
 *   Semiconductor ASA integrated circuit in a product or a software update for
 *   such product, must reproduce the above copyright notice, this list of
 *   conditions and the following disclaimer in the documentation and/or other
 *   materials provided with the distribution.
 *
 *   3. Neither the name of Nordic Semiconductor ASA nor the names of its
 *   contributors may be used to endorse or promote products derived from this
 *   software without specific prior written permission.
 *
 *   4. This software, with or without modification, must only be used with a
 *   Nordic Semiconductor ASA integrated circuit.
 *
 *   5. Any software provided in binary form under this license must not be
 *   reverse engineered, decompiled, modified and/or disassembled.
 *
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

'use strict';

import { List, Map, OrderedMap, Record } from 'immutable';
import { hackApi } from '../actions/coreActionsHack';

const ImmutableAdapterState = Record({
    instanceId: null,
    port: null,
    serialNumber: null,
    available: false,
    scanning: false,
    advertising: false,
    connecting: false,
    address: null,
    name: null,
    firmwareVerison: null,
});

const ImmutableAdapter = Record({
    state: new ImmutableAdapterState(),
    connectedDevices: Map(),
    /* Adapter sub-reducers */
    deviceDetails: null,
    serverSetup: null,
    security: null,
    isServerSetupApplied: false,
});

const ImmutableDevice = Record({
    instanceId: null,
    isConnecting: false, // Used by UI to determine visualize we are connecting
    connected: false,
    address: null,
    addressType: null,
    advType: null,
    flags: List(),
    adData: OrderedMap(),
    name: null,
    role: null,
    minConnectionInterval: null,
    maxConnectionInterval: null,
    slaveLatency: null,
    connectionSupervisionTimeout: null,
    services: List(),
    rssi: null,
    scanResponse: false,
    bonded: false,
    securityMode: null,
    securityLevel: null,
    isExpanded: false,
    ownPeriphInitiatedPairingPending: false,
});

const ImmutableProperties = Record({
    broadcast: false,
    read: false,
    writeWoResp: false,
    write: false,
    notify: false,
    indicate: false,
    authSignedWr: false,
    reliableWr: false,
    wrAux: false,
});

const ImmutableService = Record({
    instanceId: null,
    deviceInstanceId: null,
    uuid: null,
    name: null,
    handle: null,
    expanded: false,
    discoveringChildren: false,
    children: null,
});

const ImmutableCharacteristic = Record({
    instanceId: null,
    serviceInstanceId: null,
    uuid: null,
    name: null,
    declarationHandle: null,
    valueHandle: null,
    value: List(),
    properties: new ImmutableProperties(),
    readPerm: null,
    writePerm: null,
    fixedLength: false,
    maxLength: null,
    expanded: false,
    discoveringChildren: false,
    children: null,
    errorMessage: null,
});

const ImmutableDescriptor = Record({
    instanceId: null,
    characteristicInstanceId: null,
    uuid: null,
    name: null,
    handle: null,
    value: List(),
    readPerm: null,
    writePerm: null,
    fixedLength: false,
    maxLength: null,
    errorMessage: null,
});

export function getInstanceIds(instanceId) {
    const instanceIds = {
        address: null,
        device: null,
        service: null,
        characteristic: null,
        descriptor: null,
    };

    if (!instanceId) {
        return instanceIds;
    }

    const idArray = instanceId.split('.');

    if (idArray.length > 0) {
        instanceIds.address = idArray[0];
    }

    if (idArray.length > 1) {
        instanceIds.device = idArray.slice(0, 2).join('.');
    }

    if (idArray.length > 2) {
        instanceIds.service = idArray.slice(0, 3).join('.');
    }

    if (idArray.length > 3) {
        instanceIds.characteristic = idArray.slice(0, 4).join('.');
    }

    if (idArray.length > 4) {
        instanceIds.descriptor = idArray.slice(0, 5).join('.');
    }

    return instanceIds;
}

export function getImmutableAdapterState(adapterState) {
    return new ImmutableAdapterState({
        instanceId: adapterState.instanceId,
        port: adapterState.port,
        serialNumber: adapterState.serialNumber,
        available: adapterState.available,
        scanning: adapterState.scanning,
        advertising: adapterState.advertising,
        connecting: adapterState.connecting,
        address: adapterState.address,
        name: adapterState.name,
        firmwareVerison: adapterState.firmwareVerison,
    });
}

export function getImmutableAdapter(adapter) {
    return new ImmutableAdapter({
        port: adapter.state.port,
        state: getImmutableAdapterState(adapter.state),
        connectedDevices: Map(),
        /* Adapter sub-reducers */
        deviceDetails: undefined,
        serverSetup: undefined,
        security: undefined,
    });
}

export function getImmutableDevice(device) {
    return new ImmutableDevice({
        instanceId: device.instanceId,
        connected: device.connected,
        address: device.address,
        addressType: device.addressType,
        name: device.name,
        advType: device.advType,
        adData: Map(device.adData),
        flags: List(device.flags),
        role: device.role,
        minConnectionInterval: device.minConnectionInterval,
        maxConnectionInterval: device.maxConnectionInterval,
        slaveLatency: device.slaveLatency,
        connectionSupervisionTimeout: device.connectionSupervisionTimeout,
        services: List(device.services),
        rssi: device.rssi,
        scanResponse: device.scanResponse,
        ownPeriphInitiatedPairingPending: device.ownPeriphInitiatedPairingPending,
    });
}

export function getImmutableProperties(properties) {
    return new ImmutableProperties({
        broadcast: properties.broadcast,
        read: properties.read,
        writeWoResp: properties.writeWoResp || properties.write_wo_resp,
        write: properties.write,
        notify: properties.notify,
        indicate: properties.indicate,
        authSignedWr: properties.authSignedWr || properties.auth_signed_wr,
        reliableWr: properties.reliableWr || properties.reliable_wr,
        wrAux: properties.wrAux ? properties.wrAux : properties.wr_aux,
    });
}

export function getImmutableService(service) {
    return new ImmutableService({
        instanceId: service.instanceId,
        deviceInstanceId: service.deviceInstanceId,
        uuid: service.uuid,
        name: service.name,
        handle: service.startHandle,
        children: service.children,
    });
}

export function getImmutableCharacteristic(characteristic) {
    const properties = characteristic.properties || {};
    return new ImmutableCharacteristic({
        instanceId: characteristic.instanceId,
        serviceInstanceId: characteristic.serviceInstanceId,
        uuid: characteristic.uuid,
        name: characteristic.name,
        declarationHandle: characteristic.declarationHandle,
        valueHandle: characteristic.valueHandle,
        value: List(characteristic.value),
        properties: getImmutableProperties(properties),
        readPerm: characteristic.readPerm,
        writePerm: characteristic.writePerm,
        fixedLength: characteristic.fixedLength,
        maxLength: characteristic.maxLength,
        children: characteristic.children,
    });
}

function getImmutableDescriptorValue(descriptor) {
    const {
        uuid,
        value,
    } = descriptor;

    const descriptorInstanceIds = getInstanceIds(descriptor.instanceId);

    if (descriptorInstanceIds.device === 'local.server' && uuid === '2902') {
        let cccdValue = new Map();
        const keys = Object.keys(value);
        for (let i = 0; i < keys.length; i += 1) {
            const deviceInstanceId = keys[i];
            cccdValue = cccdValue.set(deviceInstanceId, List(value[deviceInstanceId]));
        }

        return cccdValue;
    }

    return List(value);
}

export function getImmutableDescriptor(descriptor) {
    const value = getImmutableDescriptorValue(descriptor);
    return new ImmutableDescriptor({
        instanceId: descriptor.instanceId,
        characteristicInstanceId: descriptor.characteristicInstanceId,
        uuid: descriptor.uuid,
        name: descriptor.name,
        handle: descriptor.handle,
        value,
        readPerm: descriptor.readPerm,
        writePerm: descriptor.writePerm,
        fixedLength: descriptor.fixedLength,
        maxLength: descriptor.maxLength,
    });
}

export default function asImmutable(mutableObject) {
    // NOTE: No idea why instanceof does not work for some types
    if (mutableObject.constructor.name === 'AdapterState') {
        return getImmutableAdapterState(mutableObject);
    } else if (mutableObject instanceof hackApi.bleDriver.Adapter) {
        return getImmutableAdapter(mutableObject);
    } else if (mutableObject instanceof hackApi.bleDriver.Device) {
        return getImmutableDevice(mutableObject);
    } else if (mutableObject instanceof hackApi.bleDriver.Service) {
        return getImmutableService(mutableObject);
    } else if (mutableObject instanceof hackApi.bleDriver.Characteristic) {
        return getImmutableCharacteristic(mutableObject);
    } else if (mutableObject instanceof hackApi.bleDriver.Descriptor) {
        return getImmutableDescriptor(mutableObject);
    }
    throw new Error('Explode!');
}
