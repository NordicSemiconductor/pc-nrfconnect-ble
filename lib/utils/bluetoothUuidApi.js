/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import { readFileSync, writeFile } from 'fs';
import path from 'path';
import { getUserDataDir, logger } from 'pc-nrfconnect-shared';

const BLUETOOTH_NUMBERS_API_DISPLAY_URL = `https://github.com/NordicSemiconductor/bluetooth-numbers-database/tree/master/v1`;
const BLUETOOTH_NUMBERS_API_URL =
    'https://raw.githubusercontent.com/NordicSemiconductor/bluetooth-numbers-database/master/v1/';
const USER_DATA_DIR = getUserDataDir();
const UUID_FILENAMES = {
    services: 'service_uuids.json',
    characteristics: 'characteristic_uuids.json',
    descriptors: 'descriptor_uuids.json',
};

function writeToFile(fileName, data) {
    const filePath = path.join(USER_DATA_DIR, fileName);
    const fileContent = JSON.stringify(data, null, 4);
    writeFile(filePath, fileContent, err => {
        if (err) {
            logger.debug(`An error ocurred creating the file ${err.message}`);
        }
    });
}

function readFromFile(fileName) {
    const filePath = path.join(USER_DATA_DIR, fileName);
    const uuidDefs = readFileSync(filePath);
    return JSON.parse(uuidDefs);
}

function getFromApi(fileName) {
    const url = BLUETOOTH_NUMBERS_API_URL + fileName;
    return fetch(url)
        .then(response => {
            if (response.status !== 200) {
                throw new Error(
                    `Response code was ${response.status} when trying to access ${url} `
                );
            }
            return response.json();
        })
        .then(data => {
            writeToFile(fileName, data);
            return data;
        })
        .catch(err => {
            throw err;
        });
}

export default () => {
    const uuidKeys = Object.keys(UUID_FILENAMES);
    const responses = uuidKeys.map(key => getFromApi(UUID_FILENAMES[key]));
    return Promise.all(responses)
        .then(([services, characteristics, descriptors]) => {
            logger.info(
                `Updated list of uuids with data from ${BLUETOOTH_NUMBERS_API_DISPLAY_URL}`
            );
            return {
                services,
                characteristics,
                descriptors,
            };
        })
        .catch(err => {
            logger.error(`${err}. Falling back to stored data`);
            try {
                const [services, characteristics, descriptors] = uuidKeys.map(
                    key => readFromFile(UUID_FILENAMES[key])
                );
                return {
                    services,
                    characteristics,
                    descriptors,
                };
            } catch (error) {
                return {};
            }
        });
};
