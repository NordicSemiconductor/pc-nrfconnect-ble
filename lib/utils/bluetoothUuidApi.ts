/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import { readFileSync, writeFile } from 'fs';
import path from 'path';
import { getUserDataDir, logger } from 'pc-nrfconnect-shared';

type Entries = {
    name: string;
    identifier: string;
    uuid: string;
    source: string;
}[];

const BLUETOOTH_NUMBERS_API_DISPLAY_URL = `https://github.com/NordicSemiconductor/bluetooth-numbers-database/tree/master/v1`;
const BLUETOOTH_NUMBERS_API_URL =
    'https://raw.githubusercontent.com/NordicSemiconductor/bluetooth-numbers-database/master/v1/';
const USER_DATA_DIR = getUserDataDir();
const UUID_FILENAMES = {
    services: 'service_uuids.json',
    characteristics: 'characteristic_uuids.json',
    descriptors: 'descriptor_uuids.json',
};

async function writeToFile(fileName: string, data: string) {
    const filePath = path.join(USER_DATA_DIR, fileName);
    const fileContent = JSON.stringify(data, null, 4);
    await writeFile(filePath, fileContent, err => {
        if (err) {
            logger.debug(`An error ocurred creating the file ${err.message}`);
        }
    });
}

function readFromFile(fileName: string): Entries {
    const filePath = path.join(USER_DATA_DIR, fileName);
    const uuidDefs = readFileSync(filePath);
    return JSON.parse(uuidDefs.toString());
}

async function getFromApi(fileName: string): Promise<Entries> {
    const url = BLUETOOTH_NUMBERS_API_URL + fileName;
    const response = await fetch(url);
    if (!response.ok) {
        throw new Error(
            `Response code was ${response.status} when trying to access ${url} `
        );
    }
    const data = await response.json();
    await writeToFile(fileName, data);
    return data;
}

export const getAllUuids = async () => {
    const uuidKeys = Object.keys(
        UUID_FILENAMES
    ) as (keyof typeof UUID_FILENAMES)[];
    const responses = uuidKeys.map(key => getFromApi(UUID_FILENAMES[key]));

    try {
        const [services, characteristics, descriptors] = await Promise.all(
            responses
        );

        logger.info(
            `Updated list of uuids with data from ${BLUETOOTH_NUMBERS_API_DISPLAY_URL}`
        );

        return {
            services,
            characteristics,
            descriptors,
        };
    } catch (err) {
        logger.error(`${err}. Falling back to stored data`);
        try {
            const [services, characteristics, descriptors] = uuidKeys.map(key =>
                readFromFile(UUID_FILENAMES[key])
            );
            return {
                services,
                characteristics,
                descriptors,
            };
        } catch {
            return {
                services: {} as Entries,
                characteristics: {} as Entries,
                descriptors: {} as Entries,
            };
        }
    }
};
