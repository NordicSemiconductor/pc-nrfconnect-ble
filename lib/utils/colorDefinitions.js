/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

'use strict';

import { logger } from 'pc-nrfconnect-shared';

import '../../resources/css/colordefinitions.scss';

const CSS_CLASS_PREFIX = 'colordefinition';
const FALLBACK_COLOR = { r: 255, g: 0, b: 0 };

function initColors(rules) {
    const colors = {};
    rules.forEach(rule => {
        const regex = new RegExp(`.${CSS_CLASS_PREFIX}-(.*)`);
        const key = regex.exec(rule.selectorText)[1];
        colors[key] = rule.style.color;
    });
    return colors;
}

function findColorDefinitionRules() {
    const rules = [];
    for (let i = 0; i < document.styleSheets.length; i += 1) {
        const { cssRules } = document.styleSheets[i];
        for (let j = 0; j < cssRules.length; j += 1) {
            const rule = cssRules[j];
            if (
                rule.selectorText &&
                rule.selectorText.indexOf(CSS_CLASS_PREFIX) !== -1
            ) {
                rules.push(rule);
            }
        }
    }
    return rules;
}

const colors = initColors(findColorDefinitionRules());

export default {
    getColor(color) {
        const rgbString = colors[color];
        if (!rgbString) {
            logger.debug(`Color ${color} is not defined`);
            return FALLBACK_COLOR;
        }
        const rgbArray = rgbString.replace(/[^\d,]/g, '').split(',');
        return {
            r: rgbArray[0],
            g: rgbArray[1],
            b: rgbArray[2],
        };
    },
};
