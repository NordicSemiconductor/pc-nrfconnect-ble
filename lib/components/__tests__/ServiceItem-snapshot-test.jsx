/*
 * Copyright (c) 2015 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

/* eslint-disable import/first */

jest.mock('../../utils/colorDefinitions', () => ({
    getColor: () => ({ r: 255, g: 255, b: 255 }),
}));
jest.mock('../../utils/uuid_definitions', () => {});

import React from 'react';
import renderer from 'react-test-renderer';

import { getImmutableService } from '../../utils/api';
import ServiceItem from '../ServiceItem';

const service = getImmutableService({
    instanceId: 'service-1',
    deviceInstanceId: 'service-deviceInstance-1',
    uuid: 'service-uuid-1',
    name: 'Service Name',
    children: [],
});

describe('ServiceItem', () => {
    it('should render correctly with no button', () => {
        const tree = renderer.create(<ServiceItem item={service} />).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render correctly with button', () => {
        const button = {
            onClick: () => {},
            icon: 'path/to/icon.png',
        };
        const tree = renderer
            .create(<ServiceItem item={service} button={button} />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });
});
