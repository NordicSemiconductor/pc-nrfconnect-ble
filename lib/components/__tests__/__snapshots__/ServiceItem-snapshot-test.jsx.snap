exports[`ServiceItem should render correctly with button 1`] = `
<div>
  <div
    className="service-item neutral-background"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "rgb(255, 255, 255)",
      }
    }>
    <div
      className="expand-area"
      onClick={[Function]}>
      <div
        className="bar1" />
      <div
        className="icon-wrap">
        <i
          className="icon-slim icon-right-dir"
          style={Object {}} />
      </div>
    </div>
    <div
      className="content-wrap">
      <div
        className="content">
        <div
          className="service-name truncate-text"
          title="UUID: service-uuid-1">
          Service Name
        </div>
        <button
          className="btn btn-primary btn-nordic"
          onClick={[Function]}>
          <img
            src="path/to/icon.png" />
        </button>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "display": "none",
      }
    } />
</div>
`;

exports[`ServiceItem should render correctly with no button 1`] = `
<div>
  <div
    className="service-item neutral-background"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "rgb(255, 255, 255)",
      }
    }>
    <div
      className="expand-area"
      onClick={[Function]}>
      <div
        className="bar1" />
      <div
        className="icon-wrap">
        <i
          className="icon-slim icon-right-dir"
          style={Object {}} />
      </div>
    </div>
    <div
      className="content-wrap">
      <div
        className="content">
        <div
          className="service-name truncate-text"
          title="UUID: service-uuid-1">
          Service Name
        </div>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "display": "none",
      }
    } />
</div>
`;
